src/basic/fileio.c

```diff
@@ -1018,10 +1018,9 @@ int chase_symlinks_and_fopen_unlocked(
         if (fd < 0)
                 return fd;
 
-        r = fdopen_unlocked(fd, open_flags, &f);
+        r = take_fdopen_unlocked(&fd, open_flags, &f);
         if (r < 0)
                 return r;
-        TAKE_FD(fd);
 
         *ret_file = f;
```

Comment:
This could be optimized further as:
```C
r = take_fdopen_unlocked(&fd, open_flags, ret_file);
if (r < 0)
     return r;
f (ret_path)
      *ret_path = TAKE_PTR(final_path);
return 0;
```